### PR TITLE
✨ Add lvcreate-options

### PR DESCRIPTION
**Pull request by lentzi90**
This is the implementation of the [lvcreate-options](https://github.com/topolvm/topolvm/blob/main/docs/proposals/lvcreate-options.md) proposal.

This PR adds a new field in the `device-class` settings, namely `lvcreate-options`. It takes a list of additional arguments to pass to `lvcreate`. This can be used to create RAID volumes for example.

Please let me know if there are things I can improve!

### CODE REVIEW

File: docs/limitations.md [hunk 1]
```diff
@@ -36,3 +36,15 @@ CSI ephemeral volumes may leave orphaned logical volumes
...
+-------------------------------------------
+
+TopoLVM does not check the `lvcreate-options` that can optionally be added to a device-class.
+Therefore it cannot take them into consideration when scheduling, or do sanity checks for them.
+It is up to the user make sure that these arguments make sense and work with the VG in question.
+For example, with `--type=raid1`, the VG must have at least 2 PVs to be able to create any LVs.
+
+Note also that the options may affect the "actual" available capacity.
+With `--type=raid1`, each LV will take up twice the normal space.
+You may want to tweak the `spare-gb` setting to avoid some issues with this.
```
Comment (satoru-takeuchi) [inline docs/limitations.md:50]: It's better to describe what happens if users set `lvcreate-options`, in detail. Or it's also OK to refer to [the design document](https://github.com/topolvm/topolvm/blob/main/docs/proposals/lvcreate-options.md#design-details)

File: docs/lvmd.md [hunk 2]
```diff
@@ -42,14 +46,15 @@ device-classes:
...
+| Name                | Type     | Default | Description                                                                        |
+| ------------------- | ------   | ------- | ---------------------------------------------------------------------------------- |
+| `name`              | string   | -       | The name of a device-class.                                                        |
+| `volume-group`      | string   | -       | The group where this device-class creates the logical volumes.                     |
+| `spare-gb`          | uint64   | `10`    | Storage capacity in GiB to be spared.                                              |
+| `default`           | bool     | `false` | A flag to indicate that this device-class is used by default.                      |
+| `stripe`            | uint     | -       | The number of stripes in the logical volume.                                       |
+| `stripe-size`       | string   | -       | The amount of data that is written to one device before moving to the next device. |
+| `lvcreate-options`  | []string | -       | Extra arguments to pass to `lvcreate`, e.g. `["--type=raid1"]`. |
```
Comment (satoru-takeuchi) [inline docs/lvmd.md:57]: Please describe what happens if users specify both `stripe` and `lvcreate-options: --stripes`.

File: lvmd/vgservice_test.go [hunk 3]
```diff
@@ -183,12 +183,12 @@ func testVGService(t *testing.T, vg *command.VolumeGroup) {
...
 	}
 
-	_, err = vg.CreateVolume("test3", 1<<30, nil, 2, "4k")
+	_, err = vg.CreateVolume("test3", 1<<30, nil, 2, "4k", nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	_, err = vg.CreateVolume("test4", 1<<30, nil, 2, "4M")
+	_, err = vg.CreateVolume("test4", 1<<30, nil, 2, "4M", nil)
```
Comment (satoru-takeuchi) [inline lvmd/vgservice_test.go:191]: Please also add a test case for `lvcreateOptions != nil` case.

**Review by satoru-takeuchi — CHANGES_REQUESTED** [review]

File: e2e/lvmd3.yaml [hunk 4]
```diff
@@ -5,5 +5,10 @@ device-classes:
     default: true
     spare-gb: 1
   - name: "hdd2"
+    volume-group: "node3-myvg2"
+    spare-gb: 1
+  - name: "raid"
     volume-group: "node3-myvg3"
     spare-gb: 1
+    lvcreate-options:
+      - "--type=raid1"
```
Comment (satoru-takeuchi) [inline e2e/lvmd3.yaml:14]: Shouldn't "hdd2" still use "node3-myvg3" and "raid" use "node3-myvg4"?

Comment (lentzi90) [reply e2e/lvmd3.yaml:14]: I'm not sure why hdd2 was using node3-myvg3 and there was no node3-myvg2 so I changed it so that hdd2 is similar to the other (lvmd1.yaml and lvmd2.yaml). Is there a reason to have this different for lvmd3?
By changing it to node3-myvg2 it was possible to simplify some things in the Makefile. [This is how it was](https://github.com/topolvm/topolvm/blob/27ca8eeeb578d4c8aa7bedd93039cc630ee54e8d/e2e/Makefile#L126-L135) and then I changed it like [this](https://github.com/Nordix/topolvm/blob/0549a9a3dce5399b1d7353fd12144261e5fa2191/e2e/Makefile#L121-L126).

Comment (satoru-takeuchi) [reply e2e/lvmd3.yaml:14]: @lentzi90 

> I'm not sure why hdd2 was using node3-myvg3 and there was no node3-myvg2 so I changed it so that hdd2 is similar to the other (lvmd1.yaml and lvmd2.yaml). Is there a reason to have this different for lvmd3?

I'm asking the reason to the original author. Please wait for a while.

File: docs/limitations.md (hunk 1 above)
Comment (lentzi90) [reply docs/limitations.md:50]: Done

File: docs/lvmd.md (hunk 2 above)
Comment (lentzi90) [reply docs/lvmd.md:57]: Done

File: docs/limitations.md [hunk 5]
```diff
@@ -36,3 +36,15 @@ CSI ephemeral volumes may leave orphaned logical volumes
 
 The logical volume created by CSI ephemeral volumes may be left behind by restarting the node.
 This problem is because the kubelet on the restarted node may fail to remove the logical volume through the CSI driver when pods are removed.
+
+Extra arguments to lvcreate are not checked
```
[the remark this answers is not in the document]
Comment (lentzi90) [reply docs/limitations.md]: Done

File: lvmd/vgservice_test.go (hunk 3 above)
Comment (lentzi90) [reply lvmd/vgservice_test.go:191]: Added, but it looks like it is always skipping these tests in the CI? I added an issue for it: #438

File: docs/limitations.md [hunk 6]
```diff
@@ -36,3 +36,49 @@ CSI ephemeral volumes may leave orphaned logical volumes
...
+Notice the following:
+
+1. The VG *must* have at least two PVs in it with enough capacity or volume creation will fail.
+   This is a requirement coming from the RAID configuration and is up to the user to take into account when creating the VG and device-class.
+2. Since the data is mirrored on the two disks, it takes up twice as much space.
+   If we ask for a volume with 1 GB capacity, it will use 1 GB on each disk, i.e. 2 GB total of the VG.
+   TopoLVM does not know about this so it will not be considered when doing scheduling decisions.
+
+To help with the scheduling we can set `spare-gb` to the size of one disk.
+For example, if `disk1` and `disk2` are 100 GB each, we can set `space-gb` to 100 GB so that the reported capacity from TopoLVM would be 100 GB.
```
Comment (cupnes) [inline docs/limitations.md:77]:
```suggestion
For example, if `disk1` and `disk2` are 100 GB each, we can set `spare-gb` to 100 GB so that the reported capacity from TopoLVM would be 100 GB.
```

File: e2e/Makefile [hunk 7]
```diff
@@ -138,9 +146,14 @@ stop-lvmd:
 	$(MAKE) shutdown-kind
 	for i in $$(seq 3); do \
 		if systemctl is-active -q lvmd$$i.service; then $(SUDO) systemctl stop lvmd$$i.service; fi; \
-		for j in $$(seq 3); do \
+		for j in $$(seq 4); do \
 			if [ -f $(BACKING_STORE)/backing_store$${i}_$${j} ]; then \
```
Comment (cupnes) [inline e2e/Makefile:150]: This if statement is no longer needed because all nodes have the same backing store files and VGs.

**Review by cupnes — CHANGES_REQUESTED** [review]

File: e2e/lvmd3.yaml (hunk 4 above)
Comment (satoru-takeuchi) [reply e2e/lvmd3.yaml:14]: @lentzi90 He said "it's by mistake", haha. So please keep this code as-is and please reflect the @cupnes 's comment. We'll merge this PR after that.

File: docs/lvmd.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lentzi90) [reply docs/lvmd.md]: Fixed

File: docs/limitations.md (hunk 6 above)
Comment (lentzi90) [reply docs/limitations.md]: Thanks, fixed!

File: e2e/Makefile (hunk 7 above)
Comment (lentzi90) [reply e2e/Makefile]: Good point! Removed it.

**Review by satoru-takeuchi — APPROVED** [review]

File: e2e/lvmd3.yaml (hunk 4 above)
Comment (lentzi90) [reply e2e/lvmd3.yaml:14]: Thank you!

**Review by cupnes — APPROVED** [review]

**Comment by satoru-takeuchi** [comment]
@lentzi90 Thank you for your contribution!
